### DISCUSSION TITLE
add toJSON() to the validator

### DISCUSSION THREAD

**Discussion by ad-on-is** [adonisjs/core#4480] [Ideas]
I'm the developer of AutoSwagger.

We're trying to implement a feature to automatically  describe a request body based on a validator.

Unfortunately we currently cannot parse the vine.object.

Would it be possible to add a `toJson()` function to the validator, that returns the object fields?

```ts
const exampleValidator = vine.object({
title: vine.string(),
data: vine.object({
foo: vine.string()
})
})

console.log(exampleValidator.toJson())

/*
{
title: "string",
data: {
foo: string
}
}

*/

```
Issue in question: https://github.com/ad-on-is/adonis-autoswagger/issues/35#issuecomment-2019528848

**Comment by thetutlage** [comment]
If I get it right, you are looking for a way to convert the schema into a JSON tree? I think, this tree has to be quite verbose and should return the validation rules as well. 

So yeah, I would need a proper proposal on which format to follow.

**Reply by ad-on-is** [reply]
Virk, that would be amazing!

The validation rules would for sure be a bonus in this case. 

As of the format, the scope of the feature is not to provide a full set of validation possibilities, since this would exceed the sheer possibilities of vinejs/adonisjs.  Just enough to provide a proper swagger description and examples. 

```ts
const exampleValidator = vine.compile(
  vine.object({
    title: vine.string().minLength(6),
    data: vine.object({
      foo: vine.string().email(),
      bar: vine.number().max(10).min(5),
      arr: vine.array(vine.number().min(10).max(20)),
      regex: vine.string().regex(/^[a-z]+$/),
      baz: vine.enum(['foo', 'bar']).nullable(),
      opts: vine.string().optional(),
    }),
  })
)
```

```ts 
// exampleValidator.toJSON()
const toJSON = {
  type: 'object',
  optional: false,
  nullable: false,
  fields: {
    title: {
      type: 'string',
      optional: false,
      nullable: false,
      rules: {
        minLength: 6,
      },
    },
    data: {
      optional: false,
      nullable: false,
      type: 'object',
      fields: {
        foo: {
          type: 'string',
          rules: { email: true },
          nullable: false,
          optional: false,
        },
        bar: {
          type: 'number',
          rules: { max: 10, min: 5 },
          nullable: false,
          optional: false,
        },
        arr: {
          type: 'array',
          items: { type: 'number', rules: { min: 10, max: 20 }, nullable: false, optional: false },
        },
        regex: {
          type: 'string',
          rules: { regex: /^[a-z]+$/ },
          nullable: false,
          optional: false,
        },
        baz: {
          type: 'enum',
          values: ['foo', 'bar'],
          nullable: true,
        },
        opts: {
          type: 'string',
          rules: {},
          optional: true,
          nullable: false,
        },
      },
    },
  },
}
```
@HasanAshab Any other suggestions?

**Reply by HasanAshab** [reply]
@ad-on-is Thats perfect!

**Comment by HasanAshab** [comment]
Also we need to access **schema** on compiled validator (VineValidator), as most of the developers export the compiled validator instead of schema (which is recommended)

**Comment by thetutlage** [comment]
I see you are basically looking for a VineJS schema to JSON schema convertor, which is a little bit time-consuming task. I have added it to my Insiders todo-list.

Meanwhile, I have added `toJSON` method which returns the compiled schema and refs from a validator. Checkout this commit to understand the return value. https://github.com/vinejs/vine/commit/52599338ec99a8320c6c3ae1993467558024e140

**Reply by ad-on-is** [reply]
This is more than perfect. Thank you! 

Just to verify, `properties` always carries the nested properties of an object (and not just the top-level), right?

From my example above, this would roughly result in something like this.

```json
{
"properties": [
  {
   "fieldName": "data", 
   "properties": [
      {"fieldName":"foo"},{"fieldName": "bar"}, {"fieldName": "baz"}
   ]
  }
]
}
```

**Reply by thetutlage** [reply]
Yes, nested fields will be represented as a tree. Moreover, checkout the compiler repo README file as well https://github.com/vinejs/compiler it has some level of documentation for the schema tree

**Reply by ad-on-is** [reply]
Excellent! For our use case this is pretty much all we need.
